File: electron/scripts/state_recover.mjs
import path from "node:path";
import { fileURLToPath } from "node:url";
import { logError, logInfo } from "./logger.mjs";
import { recoverStateFromArgs } from "./state_recover_helpers.mjs";

/**
 * Runs state recovery and prints a concise summary.
 * @param {string[]} argv - CLI arguments excluding `node` and script path.
 */
export function runStateRecover(argv) {
    const RESULT = recoverStateFromArgs(argv);
    logInfo(`Recovered source type: ${RESULT.sourceType}`);
    logInfo(`Input path: ${RESULT.inputPath}`);
    logInfo(`User data dir: ${RESULT.userDataDir}`);
    logInfo(`Books: ${RESULT.counts.books}`);
    logInfo(`Sessions: ${RESULT.counts.sessions}`);
    logInfo(`Schedule rows: ${RESULT.counts.scheduleRows}`);
    logInfo(`Schedule completions: ${RESULT.counts.scheduleCompletions}`);
    if (RESULT.backups.length > 0) {
        logInfo("Created backups:");
        for (const BACKUP_PATH of RESULT.backups) {
            logInfo(`- ${BACKUP_PATH}`);
        }
    }
}

const SCRIPT_PATH = fileURLToPath(import.meta.url);
if (path.resolve(process.argv[1] || "") === SCRIPT_PATH) {
    try {
        runStateRecover(process.argv.slice(2));
    } catch (error) {
        logError("State recovery failed.", error);
        process.exitCode = 1;
    }
}

File: apps/bartleby/src-tauri/src/bin/state_recover.rs
use std::io::{self, Write};
use std::process::ExitCode;

use bartleby_app_lib::recover_state_from_args;

fn main() -> ExitCode {
    let args = std::env::args().skip(1).collect::<Vec<_>>();
    match recover_state_from_args(&args) {
        Ok(result) => write_success(&result),
        Err(error) => {
            write_failure(&error);
            ExitCode::from(1)
        }
    }
}

fn write_failure(error: &str) {
    let _ = io::stderr().write_all(format!("State recovery failed: {error}\n").as_bytes());
}

fn write_success(result: &bartleby_app_lib::RecoverySummary) -> ExitCode {
    let _ = io::stdout().write_all(success_message(result).as_bytes());
    ExitCode::SUCCESS
}

fn success_message(result: &bartleby_app_lib::RecoverySummary) -> String {
    let mut lines = vec![
        format!("Recovered source type: {}", result.source_type),
        format!("Input path: {}", result.input_path),
        format!("User data dir: {}", result.user_data_dir),
        format!("Books: {}", result.counts.books),
        format!("Sessions: {}", result.counts.sessions),
        format!("Schedule rows: {}", result.counts.schedule_rows),
        format!(
            "Schedule completions: {}",
            result.counts.schedule_completions
        ),
    ];
    lines.extend(backup_lines(&result.backups));
    format!("{}\n", lines.join("\n"))
}

fn backup_lines(backups: &[String]) -> Vec<String> {
    if backups.is_empty() {
        return Vec::new();
    }
    let mut lines = vec!["Created backups:".to_string()];
    lines.extend(backups.iter().map(|backup_path| format!("- {backup_path}")));
    lines
}
